src/sail_json_backend/json.ml

```diff
@@ -231,20 +232,69 @@ let parse_encdec i mc format =
       end
   | _ -> assert false
 
+let filter_non_operands components =
+  let rec aux acc = function
+    | [] -> List.rev acc
+    | hd :: tl -> if String.trim hd = "spc" then tl else aux (hd :: acc) tl
+  in
+  aux [] components
+
+let extract_operands k regex components =
+  let rec aux acc = function
+    | [] -> List.rev acc
+    | hd :: tl ->
+        if Str.string_match regex hd 0 then (
+          let operand = Str.matched_group 1 hd in
+          try
+            let comma_index = String.index operand ',' in
+            debug_print ("Operand before trimming: " ^ operand);
+            let trimmed = String.sub operand 0 comma_index in
+            debug_print ("Final trimmed operand: " ^ trimmed);
+            let inputl = Hashtbl.find inputs k in
+            if List.mem trimmed inputl then aux (trimmed :: acc) tl else aux acc tl
+          with Not_found -> aux (operand :: acc) tl
+        )
+        else aux acc tl
+  in
+  aux [] components
+
+let extract_and_map_operands k =
+  let components = Hashtbl.find assembly k in
+  let regex = Str.regexp ".+(\\(.*\\))" in
+  let filtered_components = filter_non_operands components in
+  let operandl = extract_operands k regex filtered_components in
+  let opmap = List.combine (Hashtbl.find inputs k) (Hashtbl.find sigs k) in
+  let operand_with_type =
+    List.map
+      (fun op ->
+        match List.find_opt (fun (name, _) -> String.equal name op) opmap with
+        | Some (_, t) -> (op, t)
+        | None -> (op, "")
+      )
+      operandl
+  in
+  debug_print
+    ("Adding to operands hashtable: " ^ k ^ " -> "
+    ^ String.concat ", " (List.map (fun (op, t) -> Printf.sprintf "(%s, %s)" op t) operand_with_type)
+    );
+  Hashtbl.add operands k operand_with_type
+
 let add_assembly app_id p =
   let x = string_list_of_mpat p in
   begin
     debug_print ("assembly.add " ^ string_of_id app_id ^ " : " ^ List.hd x);
-    Hashtbl.add assembly (string_of_id app_id) x
+    Hashtbl.add assembly (string_of_id app_id) x;
+    extract_and_map_operands (string_of_id app_id)
```

Comment:
Let's just pass "x" to the called function, since the first thing it does is pull it back out of the Hashtbl, anyway.